### PR TITLE
Refine project structure

### PR DESCRIPTION
**Pull request by BoyaWu10**
The major changes include:
- Create DataClient project, which is responsible to fetch input FHIR data
- Create DataWriter project, which is responsible to write output data
- Create Core project, which is the engine of the project

### CODE REVIEW

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/SchedulerRegistrationExtensions.cs [hunk 1]
```diff
@@ -4,10 +4,13 @@
...
+using Microsoft.Health.Fhir.Synapse.Core.DataProcessor.Json;
+using Microsoft.Health.Fhir.Synapse.Core.DataProcessor.Parquet;
+using Microsoft.Health.Fhir.Synapse.Core.DataWriter;
+using Microsoft.Health.Fhir.Synapse.Core.Jobs;
+using Microsoft.Health.Fhir.Synapse.Core.Tasks;
 
-namespace Microsoft.Health.Fhir.Synapse.Scheduler
+namespace Microsoft.Health.Fhir.Synapse.Core
 {
     public static class SchedulerRegistrationExtensions
```
[the remark this answers is not in the document]
Comment (BoyaWu10) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/SchedulerRegistrationExtensions.cs]: Sure. It is renamed in the update.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.DataClient/Api/FhirApiDataClient.cs [hunk 2]
```diff
@@ -55,21 +50,29 @@ public sealed class FhirDataClient : IFhirDataClient
             _httpClient.Timeout = Timeout.InfiniteTimeSpan;
         }
 
-        public async Task<FhirElementBatchData> GetAsync(TaskContext context, CancellationToken cancellationToken = default)
+        public async Task<FhirElementBatchData> GetAsync(
+            string resourceType,
```
Comment (moria97) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.DataClient/Api/FhirApiDataClient.cs:54]: Shall we wrap the parameters into some `FhirSearchInput` class? Then we don't need to change the interface if we  change/add any search parameters in the future

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/DataProcessor/ParquetDataProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        private readonly ArrowConfiguration _arrowConfiguration;
+        private readonly ILogger<ParquetDataProcessor> _logger;
+        private readonly ParquetConverterWrapper _parquetConverterWrapper;
+
+        public ParquetDataProcessor(
+            IFhirSchemaManager<FhirParquetSchemaNode> fhirSchemaManager,
+            IOptions<ArrowConfiguration> arrowConfiguration,
+            ILogger<ParquetDataProcessor> logger)
+        {
+            EnsureArg.IsNotNull(fhirSchemaManager);
```
[the remark this answers is not in the document]
Comment (BoyaWu10) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/DataProcessor/ParquetDataProcessor.cs]: Thanks, fixed in the update.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/DataProcessor/ParquetDataProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+            if (schema == null)
+            {
+                _logger.LogError($"The FHIR schema node could not be found for resource type '{context.ResourceType}'.");
+                throw new ParquetDataProcessorException($"The FHIR schema node could not be found for resource type '{context.ResourceType}'.");
+            }
+
+            var inputStream = ConvertJsonDataToStream(context.ResourceType, preprocessedData.Values);
+            if (inputStream == null)
+            {
+                // Return null if no data been converted.
```
[the remark this answers is not in the document]
Comment (BoyaWu10) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/DataProcessor/ParquetDataProcessor.cs]: Thanks, fixed in the update.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/JobManager.cs [hunk 5]
```diff
@@ -58,7 +52,7 @@ public class JobManager : IDisposable, IAsyncDisposable
         /// </summary>
         /// <param name="cancellationToken">cancellation token.</param>
         /// <returns>Completed task.</returns>
-        public async Task TriggerJobAsync(CancellationToken cancellationToken = default)
+        public async Task Run(CancellationToken cancellationToken = default)
```
[the remark this answers is not in the document]
Comment (BoyaWu10) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/JobManager.cs]: Thanks, fixed in the update.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.DataClient/Api/FhirApiDataClient.cs (hunk 2 above)
Comment (BoyaWu10) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.DataClient/Api/FhirApiDataClient.cs]: Good idea. I've added a `FhirSearchParameters` class to hold all the search parameters.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/DataProcessor/ParquetDataProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+            var processedJsonData = inputData.Values
+                .Select(json => ProcessStructObject(json, schema))
+                .Where(processedResult => processedResult != null);
+
+            context.SkippedCount += inputData.Values.Count() - processedJsonData.Count();
+
+            return new JsonBatchData(processedJsonData);
+        }
+
+        private JObject ProcessStructObject(JToken structItem, FhirParquetSchemaNode schemaNode)
```
[the remark this answers is not in the document]
Comment (BoyaWu10) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/DataProcessor/ParquetDataProcessor.cs:125]: My understanding is that these JObject processing logics are designed for Parquet and may not be shared by other data formats like CDM, so currently they serve as private functions of `ParquetDataProcessor`. When these functions are used in other places or could be made public, we can create a utility class to hold them at that time.

**Review by moria97 — APPROVED** [review]

**Pull request merged**